### ISSUE TITLE
Getting Yul ASTs requires asking for ast at contract level instead of at file level

### ISSUE THREAD

**Issue by haltman-at** [ethereum/solidity#14452]
## Description

When compiling in Yul mode, including `"ast"` in the file-level output selection doesn't result in getting ASTs like it does when compiling Solidity.  Instead you have to set it at the contract level.  This makes no sense, because ASTs are a file-level matter, and is also inconsistent with how the compiler acts when compiling Solidity.

## Environment

- Compiler version: 0.8.21
- Target EVM version (as per compiler settings): default

## Steps to Reproduce

Really any Yul code will do here!  What's relevant here is the output selection.  An `outputSelection` of

```
{
  "*": {
    "": ["ast"],
    "*": [
      "evm.bytecode.object"
    ]
  }
}
```

won't get the AST,

whereas one of

```

{
  "*": {
    "": [],
    "*": [
      "ast",
      "evm.bytecode.object"
    ]
  }
}
```

will.

**Comment by r0qs** [comment]
Thanks @haltman-at, yeah, the Yul is generated per contract and, currently, only one Yul object can be compiled at the same time (so the ID would be always the same), whereas the solidity AST is cross contract. We may need to change that indeed and include `ast` in the file-level for Yul as well.

**Comment by cameel** [comment]
Yeah, this is definitely a bug in #14177. In Yul there's always one contract per file, so you could technically associate the AST with a contract, but it's really the source unit that has an AST. The original PR confusingly made it a contract-level output so [I pointed it out back then and it was changed](https://github.com/ethereum/solidity/pull/14177#discussion_r1194079448), but looks like this was only done for the output JSON.

Note that this was really meant to be an experimental feature (even though it looks like we did not make that very clear and even included a changelog entry for it). As such it wasn't very thoroughly reviewed or tested. The intention was to merge it quickly. If anything else about it is off, please do report it :)
